File: day2_factory/guru/ts/creator.ts
import * as pdt from "./product"

// The Creator class declares the factory method that is supposed to return an
// object of a Product class. The Creator's subclasses usually provide the
// implementation of this method.

export abstract class Creator {
    // Note that the Creator may also provide some default implementation of the
    // factory method.
    public abstract factoryMethod(): pdt.Product;

    // Also note that, despite its name, the Creator's primary responsibility is
    // not creating products. Usually, it contains some core business logic that
    // relies on Product objects, returned by the factory method. Subclasses can
    // indirectly change that business logic by overriding the factory method
    // and returning a different type of product from it.
    public someOperation(): string {
        // Call the factory method to create a Product object.
        const product = this.factoryMethod();

        // Now, use the product.
        return `Creator: The same creator's code has just worked with ${product.operation()}`;
    }
}

// Concrete Creators override the factory method in order to change the
// resulting product's type.
export class ConcreteCreator1 extends Creator {
    // Note that the signature of the method still uses the abstract product
    // type, even though the concrete product is actually returned from the
    // method. This way the Creator can stay independent of concrete product
    // classes.
    public factoryMethod(): pdt.Product {
        return new pdt.ConcreteProduct1();
    }
}

export class ConcreteCreator2 extends Creator {
    public factoryMethod(): pdt.Product {
        return new pdt.ConcreteProduct2();
    }
}

File: day02_factory/guru/py/creator.py
from abc import ABC, abstractmethod
from overrides import overrides

from product import *


class Creator(ABC):
    """
    The Creator class declares the factory method that is supposed to return an
    object of a Product class. The Creator's subclasses usually provide the
    implementation of this method.
    """

    @abstractmethod
    def factory_method(self):
        """
        Note that the Creator may also provide some default implementation of
        the factory method.
        """
        pass

    def some_operation(self) -> str:
        """
        Also note that, despite its name, the Creator's primary responsibility
        is not creating products. Usually, it contains some core business logic
        that relies on Product objects, returned by the factory method.
        Subclasses can indirectly change that business logic by overriding the
        factory method and returning a different type of product from it.
        """

        # Call the factory method to create a Product object.
        product: Product = self.factory_method()

        # Now, use the product.
        result = f'Creator: The same creator\'s code has just worked with {product.operation()}'

        return result


"""
Concrete Creators override the factory method in order to change the resulting
product's type.
"""
class ConcreteCreator1(Creator):
    """
    Note that the signature of the method still uses the abstract product type,
    even though the concrete product is actually returned from the method. This
    way the Creator can stay independent of concrete product classes.
    """

    @overrides
    def factory_method(self) -> Product:
        return ConcreteProduct1()


class ConcreteCreator2(Creator):
    @overrides
    def factory_method(self) -> Product:
        return ConcreteProduct1()
